### PR TITLE
VReplication: Estimate lag when workflow fully throttled

### PR DESCRIPTION
**Pull request by mattlord**
## Description

We did not properly calculate the vreplication lag for a workflow when it was fully throttled — on the source/vstreamer side or the target/vplayer side.

This PR ensures that we account for these scenarios and estimate the vreplication lag for the workflow when they occur.

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/16575

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Review by shlomi-noach — APPROVED** [review]
While this PR is still in Draft, it looks good to me. Thank you!

File: go/vt/vtctl/workflow/materializer.go [hunk 1]
```diff
@@ -618,7 +618,7 @@ func primaryVindexesDiffer(ms *vtctldatapb.MaterializeSettings, source, target *
 		// If neither source nor target have any vindexes, treat the answer to
 		// the question as trivially false.
 		if len(sColumnVindexes) == 0 && len(tColumnVindexes) == 0 {
-			return true
+			return false
```
Comment (mattlord) [inline go/vt/vtctl/workflow/materializer.go:621]: This is an unrelated fix for https://github.com/vitessio/vitess/issues/16565 

This had no practical impact but was clearly wrong.

Comment (deepthi) [reply go/vt/vtctl/workflow/materializer.go:621]: In general I'm not in favor of mixing two separate bug fixes into one PR, mostly because different issues have different levels of severity, and you might not want to backport both into release branches.

File: go/vt/vtctl/workflow/materializer.go [hunk 2]
```diff
@@ -618,7 +618,7 @@ func primaryVindexesDiffer(ms *vtctldatapb.MaterializeSettings, source, target *
 		// If neither source nor target have any vindexes, treat the answer to
 		// the question as trivially false.
 		if len(sColumnVindexes) == 0 && len(tColumnVindexes) == 0 {
```
Comment (deepthi) [inline go/vt/vtctl/workflow/materializer.go:620]: If you are touching this anyway, it seems better to move this up, and make it the first condition we check for.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go [hunk 3]
```diff
@@ -305,24 +304,22 @@ func (vs *vstreamer) parseEvents(ctx context.Context, events <-chan mysql.Binlog
...
 				select {
 				case <-ctx.Done():
 					return
 				default:
-					// do nothing special
+					// Do nothing special.
 				}
-				throttledHeartbeatsRateLimiter.Do(func() error {
-					return injectHeartbeat(true, checkResult.Summary())
-				})
```
Comment (mattlord) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:323]: I realized that we were getting double the heartbeats as we send them here and via the timer. I consolidated them into the main loop with the timer. Otherwise we needed to do twice the work as the other heartbeats would get interspersed with these and those would indicate that we weren't throttled when in fact we were.

File: go/vt/vttablet/tabletmanager/vreplication/vplayer.go [hunk 4]
```diff
@@ -476,33 +476,34 @@ func (vp *vplayer) recordHeartbeat() error {
...
+		vp.vr.stats.VReplicationLags.Add(strconv.Itoa(int(vp.vr.id)), time.Duration(behind/1e9)*time.Second)
+	}
+
 	// If we're not running, set ReplicationLagSeconds to be very high.
 	// TODO(sougou): if we also stored the time of the last event, we
 	// can estimate this value more accurately.
 	defer vp.vr.stats.ReplicationLagSeconds.Store(math.MaxInt64)
 	defer vp.vr.stats.VReplicationLags.Add(strconv.Itoa(int(vp.vr.id)), math.MaxInt64)
-	var sbm int64 = -1
+	var sbm int64
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletmanager/vreplication/vplayer.go:490]: What does "sbm" stand for?

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go [hunk 5]
```diff
@@ -305,24 +304,22 @@ func (vs *vstreamer) parseEvents(ctx context.Context, events <-chan mysql.Binlog
...
-				// we won't process events, until we're no longer throttling
-				logger.Infof("throttled.")
+				logger.Infof("vstreamer throttled%s: %s.", wfNameLog, checkResult.Summary())
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:322]: Would this spam the logs when throttled?

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go [hunk 6]
```diff
@@ -394,7 +391,8 @@ func (vs *vstreamer) parseEvents(ctx context.Context, events <-chan mysql.Binlog
 		case <-ctx.Done():
 			return nil
 		case <-hbTimer.C:
-			if err := injectHeartbeat(false, ""); err != nil {
+			checkResult, ok := vs.vse.throttlerClient.ThrottleCheckOKOrWaitAppName(ctx, vs.throttlerApp)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:394]: Could you please explain what this throttler check is for?

**Review by shlomi-noach — COMMENTED** [review]
I'm not sure I understand some of the changes in the last force-push, asking for clarifications inline.

File: go/vt/vttablet/tabletmanager/vreplication/vplayer.go (hunk 4 above)
Comment (mattlord) [reply go/vt/vttablet/tabletmanager/vreplication/vplayer.go:490]: I didn't name it, but I'm assuming it's an initialism for "seconds behind master". It's not obvious to the reader, so I'll rename it to `lagSecs` since I'm here.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 5 above)
Comment (mattlord) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:322]: It's a throttled logger that logs the message at most once per minute. It was added in https://github.com/vitessio/vitess/pull/14936. I thought about removing these altogether as the value was not clear to me, but instead I at least made this one much more useful IMO as it would state the component and the workflow.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 6 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:394]: In that case I encourage you to use `vs.vse.throttlerClient.ThrottleCheckOK` as opposed to `vs.vse.throttlerClient.ThrottleCheckOKOrWaitAppName`. You're not interested in the extra sleep.

This makes sense now, thank you for clarifying.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:322]: OK cool!

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 6 above)
Comment (mattlord) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go]: Ah, I see. Thank you! 🙇 That makes total sense — no need for the sleep here as it's not done in a tight loop. I should have at least checked the function comments. 🙂 I've made this change.

**Review by rohit-nayak-ps — APPROVED** [review]
Nice catch and fix.
